File: typescript/src/process1/activities.ts
import { Context } from '@temporalio/activity';
import type { RegisterRequest } from './models';

const users = new Set<string>();
const ops = new Set<string>();

const defaultSleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
let sleep = defaultSleep;

export function setActivitySleepForTest(fn: typeof sleep): void {
  sleep = fn;
}

export function resetActivityStateForTest(): void {
  users.clear();
  ops.clear();
  sleep = defaultSleep;
}

export async function createAccount(request: RegisterRequest): Promise<void> {
  if (users.has(request.userId)) {
    return;
  }
  users.add(request.userId);
  await sleep(5000);
  console.log(`Created account for ${request.userId} (${request.email})`);
}

export async function addPoints(
  userId: string,
  points: number,
  opKey: string
): Promise<void> {
  const dedupeKey = `POINTS#${userId}#${opKey}`;
  const { attempt } = Context.current().info;

  if (ops.has(dedupeKey)) {
    console.log(`[idempotent] addPoints skipped for ${userId} [op=${opKey}]`);
    return;
  }

  if (attempt <= 3) {
    throw new Error(`DB connection failed (simulated), attempt ${attempt}`);
  }

  await sleep(5000);
  console.log(`Added ${points} points to ${userId} [op=${opKey}]`);
  ops.add(dedupeKey);
}

export async function sendEmail(email: string, messageId: string): Promise<void> {
  const dedupeKey = `EMAIL#${messageId}`;
  if (ops.has(dedupeKey)) {
    console.log(`[idempotent] sendEmail skipped for ${email} [msgId=${messageId}]`);
    return;
  }
  ops.add(dedupeKey);
  await sleep(5000);
  console.log(`Sent welcome email to ${email} [msgId=${messageId}]`);
}

File: python/src/temporal_learn/process1/activities.py
import asyncio

from temporalio import activity

from .models import RegisterRequest

# Demo-only：用 memory set 模擬資料庫 unique constraint，方便展示 idempotency。
# 生產環境不可依賴 worker memory，應改用 DB / Redis / 外部服務的 idempotency key。
_demo_users: set[str] = set()
_demo_ops: set[str] = set()


def reset_activity_state_for_test() -> None:
    _demo_users.clear()
    _demo_ops.clear()


# -------------------------------------------------------------------------
# Activity: create_account
@activity.defn
async def create_account(request: RegisterRequest) -> None:
    if request.user_id in _demo_users:
        return
    _demo_users.add(request.user_id)
    await asyncio.sleep(5)
    print(f"Created account for {request.user_id} ({request.email})")


# -------------------------------------------------------------------------
# Activity: add_points
@activity.defn
async def add_points(user_id: str, points: int, op_key: str) -> None:
    # namespace#entity#operation 的複合鍵設計，確保不同操作的 key 不會碰撞
    dedupe_key = f"POINTS#{user_id}#{op_key}"
    # 冪等性核心：相同 op_key 重複呼叫無副作用地直接返回
    if dedupe_key in _demo_ops:
        print(f"[idempotent] addPoints skipped for {user_id} [op={op_key}]")
        return

    await asyncio.sleep(5)
    print(f"Added {points} points to {user_id} [op={op_key}]")

    _demo_ops.add(dedupe_key)


# -------------------------------------------------------------------------
# Activity: send_email
@activity.defn
async def send_email(email: str, message_id: str) -> None:
    dedupe_key = f"EMAIL#{message_id}"
    if dedupe_key in _demo_ops:
        print(f"[idempotent] sendEmail skipped for {email} [msgId={message_id}]")
        return
    _demo_ops.add(dedupe_key)
    await asyncio.sleep(5)
    print(f"Sent welcome email to {email} [msgId={message_id}]")
